packages/php-wasm/web-service-worker/src/fetch-caching.ts

```diff
@@ -0,0 +1,69 @@
+import { logger } from '@php-wasm/logger';
+
+const wpCacheKey = 'playground-cache';
+
+const isCacheFull = async () => {
+	if (navigator.storage && navigator.storage.estimate) {
+		const estimatedStorage = await navigator.storage.estimate();
+		if (estimatedStorage.usage && estimatedStorage.quota) {
+			return estimatedStorage.usage >= estimatedStorage.quota;
+		}
+	}
+	return false;
+};
+
+export const addCache = async (key: string, response: Response) => {
+	if (await isCacheFull()) {
+		logger.warn('Cache is full, not caching response');
+		return;
+	}
+	try {
+		const clonedResponse = response.clone();
+		const cache = await caches.open(wpCacheKey);
+		await cache.put(key, clonedResponse);
+	} catch (e) {
+		logger.warn('Failed to cache response', e);
+	}
+};
+
+export const getCache = async (key: string) => {
+	const cache = caches.open(wpCacheKey);
+	return await cache.then((c) => c.match(key));
+};
+
+const validHostnames = ['playground.wordpress.net', '127.0.0.1', 'localhost'];
```

Comment:
Wow, I had no idea Service Workers also receive fetch events for requests to third party resources. But it appears to be true.

It looks like `fetchEvent.currentTarget.registration.scope` is available, so we could filter based on that rather than a specific set of hostnames.

[`fetchEvent.currentTarget`](https://dom.spec.whatwg.org/#ref-for-dom-event-currenttarget②) points to the ServiceWorkerGlobalScope. ServiceWorkerGlobalScope exposes the ServiceWorkerRegistration as its [registration property](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/registration), and ServiceWorkerRegistration has the [scope property](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/scope).